File: src/common/framework/display_object.c
#include "display_object.h"
#include "quad.h"


void DisplayObject_init(DisplayObject *d){
  d->parent = 0;
  DisplayObject_setTransform(d, 0, 0, 1, 1, 0, 0, 0, 0, 0);
}

void DisplayObject_setTransform(DisplayObject *d, double x, double y, double scaleX, double scaleY, double rot, double skewX, double skewY, double regX, double regY){
  d->x = x;
  d->y = y;
  d->scaleX = scaleX;
  d->scaleY = scaleY;
  d->rotation = rot;
  d->skewX = skewX;
  d->skewY = skewY;
  d->regX = regX;
  d->regY = regY;
}


void DisplayObject_draw(DisplayObject *d){
  Matrix2 m;
  DisplayObject_getConcatenatedMatrix(d, &m);
  Matrix2_append(&m, d->tex->width, 0, 0, d->tex->height, 0, 0);
  quadDrawTex(d->tex, &m);
}


void DisplayObject_getConcatenatedMatrix(DisplayObject *d, Matrix2 *outMat){
  DisplayObject *o = d;
  Matrix2_identity(outMat);
  while (o){
    Matrix2_prependTransform(outMat, o->x, o->y, o->scaleX, o->scaleY, o->rotation, o->skewX, o->skewY, o->regX, o->regY);
    o = o->parent;
  }
}

File: src/common/framework/display_object.nim
import 
  matrix2, texture

type 
  DisplayObject* {.importc: "DisplayObject", header: "display_object.h".} = object 
  
  DisplayObject* {.importc: "DisplayObject", header: "display_object.h".} = object 
    x* {.importc: "x".}: cdouble
    y* {.importc: "y".}: cdouble
    skewX* {.importc: "skewX".}: cdouble
    skewY* {.importc: "skewY".}: cdouble
    regX* {.importc: "regX".}: cdouble
    regY* {.importc: "regY".}: cdouble
    rotation* {.importc: "rotation".}: cdouble
    scaleX* {.importc: "scaleX".}: cdouble
    scaleY* {.importc: "scaleY".}: cdouble
    alpha* {.importc: "alpha".}: cdouble
    parent* {.importc: "parent".}: ptr DisplayObject
    tex* {.importc: "tex".}: ptr Texture


proc DisplayObject_init*(d: ptr DisplayObject) {.importc: "DisplayObject_init", 
    header: "display_object.h".}
proc DisplayObject_setTransform*(d: ptr DisplayObject; x: cdouble; y: cdouble; 
                                 scaleX: cdouble; scaleY: cdouble; rot: cdouble; 
                                 skewX: cdouble; skewY: cdouble; regX: cdouble; 
                                 regY: cdouble) {.
    importc: "DisplayObject_setTransform", header: "display_object.h".}
proc DisplayObject_draw*(d: ptr DisplayObject) {.importc: "DisplayObject_draw", 
    header: "display_object.h".}
proc DisplayObject_getConcatenatedMatrix*(d: ptr DisplayObject; 
    outMat: ptr Matrix2) {.importc: "DisplayObject_getConcatenatedMatrix", 
                           header: "display_object.h".}
proc DisplayObject_getConcatAlpha*(d: ptr DisplayObject): cdouble {.
    importc: "DisplayObject_getConcatAlpha", header: "display_object.h".}

